### PR TITLE
refactor: rewrite Affix CC => FC

### LINKED ISSUE #22683: Antd Affix: Memory leak error
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[Edit on CodeSandbox](https://codesandbox.io/s/wonderful-shamir-tn0fu)

### Steps to reproduce
1. Scroll a little till the Affix becomes sticky.
2. Click on Toggle button.
3. First Affix is removed and second Affix gets mounted.
4. Warning seen on console - Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
5. Note: This is particularly observed in Chrome - Emulation Mode (Responsive Viewport Mode)

### What is expected?
Warning should not appear

### What is actually happening?
Memory leak warning is seen

| Environment | Info |
|---|---|
| antd | 3.26.13 |
| React | 16.3 |
| System | Windows 10 |
| Browser | Chrome - Emulation Mode (Responsive Viewport Mode) |

### LINKED ISSUE #39892: refactor Affix to function component
https://github.com/ant-design/ant-design/blob/dc77959e7ced3f0ae393116e2f3d18cf9b3e0f14/components/affix/index.tsx#L57

### PR DESCRIPTION
**Pull request by li-jia-nan**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [x] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

- ref #39900
- close #39892
- close #44408
- close #22683

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | refactor: rewrite Affix CC => FC |
| 🇨🇳 Chinese | 重构为函数组件 |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 5a4959b</samp>

Refactored the `Affix` component to use React hooks and simplified its props and style. Adjusted the import order of types in `components/input-number/style/index.ts` for code style consistency.

### <samp>🤖 Generated by Copilot at 5a4959b</samp>

*  Refactor `Affix` component from class to functional using hooks ([link](https://github.com/ant-design/ant-design/pull/42674/files?diff=unified&w=0#diff-18949a009c0a2f7f47a0136b5808eade7a157b3d78661edc0f75463594e2ad5dL4-R4), [link](https://github.com/ant-design/ant-design/pull/42674/files?diff=unified&w=0#diff-18949a009c0a2f7f47a0136b5808eade7a157b3d78661edc0f75463594e2ad5dL42-L45), [link](https://github.com/ant-design/ant-design/pull/42674/files?diff=unified&w=0#diff-18949a009c0a2f7f47a0136b5808eade7a157b3d78661edc0f75463594e2ad5dL59-R201))
* Adjust import order in `components/input-number/style/index.ts` to follow convention of importing types after other imports ([link](https://github.com/ant-design/ant-design/pull/42674/files?diff=unified&w=0#diff-63d0bbc4ed8cb1745da77de552d75aa0fd581ef3eb2d25d0108b2d55eb5f1857L12-R15))

### CODE REVIEW

**Comment by afc163** [comment]
feature 分支吧

**Comment by li-jia-nan** [comment]
> feature 分支吧

OK，我改一下

File: components/_util/throttleByAnimationFrame.ts [hunk 1]
```diff
@@ -12,7 +8,7 @@ function throttleByAnimationFrame<T extends any[]>(fn: (...args: T) => void) {
     fn(...args);
   };
 
-  const throttled: throttledFn & throttledCancelFn = (...args: T) => {
+  const throttled = (...args: T) => {
```
Comment (li-jia-nan) [inline components/_util/throttleByAnimationFrame.ts:11]: 这里不需要声明类型，让 TS 自动推导即可

File: components/affix/index.tsx [hunk 2]
```diff
@@ -41,134 +40,65 @@ export interface AffixProps {
...
+    rootClassName,
+    children,
+    target,
+    onChange,
+  } = props;
 
-    if (target !== undefined) {
-      return target;
-    }
+  const [lastAffix, setLastAffix] = React.useState<boolean>(false);
```
Comment (kiner-tang) [inline components/affix/index.tsx:76]: 简单类型没必要传泛型，直接根据传入的初始值推导就好了

File: components/affix/index.tsx [hunk 3]
```diff
@@ -260,49 +173,86 @@ class InternalAffix extends React.Component<InternalAffixProps, AffixState> {
...
+    'target',
+    'onChange',
+    'affixPrefixCls',
+    'rootClassName',
+  ]);
+
+  return (
+    <ResizeObserver onResize={updatePosition}>
+      <div style={style} ref={placeholderNodeRef} {...otherProps}>
+        {affixStyle ? <div style={placeholderStyle} aria-hidden="true" /> : null}
```
Comment (kiner-tang) [inline components/affix/index.tsx:244]:
```suggestion
        {affixStyle && <div style={placeholderStyle} aria-hidden="true" />}
```

**Review by kiner-tang — APPROVED** [review]

File: components/affix/index.tsx (hunk 2 above)
Comment (li-jia-nan) [reply components/affix/index.tsx]: done~

**Comment by afc163** [comment]
看下覆盖率。

File: components/affix/index.zh-CN.md [hunk 4]
```diff
@@ -19,6 +19,10 @@ group:
 
 页面可视范围过小时，慎用此功能以免遮挡页面内容。
 
+> 开发者注意事项：
+>
+> 自 `5.10.0` 起，由于 Affix 组件由 class 重构为 FC，之前获取 `ref` 并调用内部实例方法的写法都会失效
```
Comment (afc163) [inline components/affix/index.zh-CN.md:24]:
```suggestion
> 自 `5.10.0` 起，由于 Affix 组件由 class 重构为 FC，之前获取 `ref` 并调用内部实例方法的写法都会失效。
```

**Comment by li-jia-nan** [comment]
> 看下覆盖率。

补上了大佬

File: components/affix/index.tsx [hunk 5]
```diff
@@ -260,49 +177,92 @@ class InternalAffix extends React.Component<InternalAffixProps, AffixState> {
...
+    'prefixCls',
+    'offsetTop',
+    'offsetBottom',
+    'target',
+    'onChange',
+    'affixPrefixCls',
+    'rootClassName',
+  ]);
+
+  if (process.env.NODE_ENV === 'test') {
```
Comment (afc163) [inline components/affix/index.tsx:245]: 为了测试覆盖率加这个很奇怪，没办法通过用例本身覆盖么？

Comment (li-jia-nan) [reply components/affix/index.tsx:245]: 这个场景不好模拟出来，这个写法就是 master 现有的写法，之前是豆酱大佬写的，我重构的时候本来去掉了，但是发现去掉之后覆盖率会掉，没办法又加回来了

Comment (afc163) [reply components/affix/index.tsx:245]: 先合了。

Comment (afc163) [reply components/affix/index.tsx:245]: https://preview-42674-ant-design.surge.sh/components/affix-cn 这个例子要顺手改下，现在看不到固定效果。

图片

**Comment by li-jia-nan** [comment]
data-picker 快照挂了，需要合一下master

File: components/affix/index.tsx (hunk 5 above)
Comment (li-jia-nan) [reply components/affix/index.tsx:259]: 改好了，加了个固定宽高

**Comment by afc163** [comment]
终于有 PR 是减 bundlesize 的了~

**Pull request merged**